File: examples/sockets.c
#include <stdlib.h> /* abort */
#include <string.h> /* memset */
#include <stdio.h> /* printf, perror */
#include <sys/types.h>

#ifdef WIN32

//#include <winsock.h> /* send, recv */
#include <winsock2.h>
#include <process.h> /* _exit */
typedef int ssize_t;

#else

#include <sys/socket.h>   /* socket(), bind(), listen(), accept() */
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <unistd.h> /* fork, write, close */
#define SOCKET int
#define INVALID_SOCKET (-1)

#endif

#include <stdbool.h>
#include "stringBuffers.h"
#include "sockets.h"

void print_socket_error_message(char *api)
{
#ifdef WIN32
    printf("Socket API call '%s' failed: error code %d\n", api, WSAGetLastError());
#else
    perror(api);
#endif
}

SOCKET socket_create_and_listen(int port)
{
    SOCKET serverSocket = 0;
    struct sockaddr_in serverName = { 0 };
    int status = 0;

#ifdef WIN32
    {
        WSADATA windowsSocketsApiData;
        WSAStartup(MAKEWORD(2, 0), &windowsSocketsApiData);
    }
#endif

    serverSocket = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (INVALID_SOCKET == serverSocket)
    {
        print_socket_error_message("socket()");
        abort();
    }
    
    serverName.sin_addr.s_addr=htonl(INADDR_ANY);
    serverName.sin_family = AF_INET;
    /* network-order */
    serverName.sin_port = htons(port);

    status = bind(serverSocket, (struct sockaddr *) &serverName, sizeof(serverName));
    if (INVALID_SOCKET == status)
    {
        print_socket_error_message("bind()");
        abort();
    }
    
    status = listen(serverSocket, 5);  // Allow 5 pending incoming connection requests
    if (INVALID_SOCKET == status)
    {
        print_socket_error_message("listen()");
        abort();
    }

    return serverSocket;
}

SOCKET socket_accept(SOCKET serverSocket)
{
    struct sockaddr_in clientName = { 0 };
    SOCKET slaveSocket;
    int clientLength = sizeof(clientName);

    (void) memset(&clientName, 0, sizeof(clientName));

    slaveSocket = accept(serverSocket, (struct sockaddr *) &clientName, &clientLength);
    if (INVALID_SOCKET == slaveSocket)
    {
        print_socket_error_message("accept()");
        abort();
    }
    
    return slaveSocket;
}

SOCKET socket_create(int port)
{
    SOCKET lsocket;
    SOCKADDR_IN lSockAddr;
    SOCKET serverSocket = 0;
    int status = 0;

#ifdef WIN32
    {
        WSADATA windowsSocketsApiData;
        WSAStartup(MAKEWORD(2, 0), &windowsSocketsApiData);
    }
#endif

    lsocket = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (INVALID_SOCKET == lsocket)
    {
        print_socket_error_message("socket()");
        abort();
    }

    memset(&lSockAddr,0, sizeof(lSockAddr));
    lSockAddr.sin_family = AF_INET;
    lSockAddr.sin_port = htons(port);
    lSockAddr.sin_addr.s_addr = inet_addr("127.0.0.1");
    status = connect(lsocket,(SOCKADDR *)&lSockAddr,sizeof(SOCKADDR_IN));
    if(status != 0)
    {
        print_socket_error_message("connect()");
        abort();
    }
    return lsocket;
}

struct server_socket {
    int handle;
};

struct server_socket *create_server_socket(int port)
{
    int handle = socket_create_and_listen(port);
    struct server_socket *serverSocket = malloc(sizeof(struct server_socket));
    serverSocket->handle = handle;
    return serverSocket;
}

#define READER_BUFFER_SIZE 4096

struct reader {
    int handle;
    char *bufferStart;
    char *bufferEnd;
    char buffer[READER_BUFFER_SIZE];
};

bool reader_read_line(struct reader *reader, struct string_buffer *buffer)
{
    string_buffer_clear(buffer);
    for (;;)
    {
        void *newline = memchr(reader->bufferStart, '\n', reader->bufferEnd - reader->bufferStart);
        if (newline != 0) {
            char *end = (char *)newline;
            if (reader->bufferStart < end && *(end - 1) == '\r')
                end--;
            string_buffer_append_chars(buffer, reader->bufferStart, end - reader->bufferStart);
            reader->bufferStart = (char *)newline + 1;
            return false;
        } else {
            string_buffer_append_chars(buffer, reader->bufferStart, reader->bufferEnd - reader->bufferStart);
            reader->bufferStart = reader->buffer;
            {
                ssize_t count = recv(reader->handle, reader->buffer, READER_BUFFER_SIZE, 0);
                if (count < 0) {
                    print_socket_error_message("recv()");
                    abort();
                }
                if (count == 0)
                    return true;
                reader->bufferEnd = reader->buffer + count;
            }
        }
    }
}

struct writer {
    int handle;
};

void writer_write_string(struct writer *writer, char *text)
{
    size_t length = strlen(text);
    send(writer->handle, text, length, 0);
}

void writer_write_string_buffer(struct writer *writer, struct string_buffer *buffer)
{
    send(writer->handle, string_buffer_get_chars(buffer), string_buffer_get_length(buffer), 0);
}

struct socket {
    int handle;
    struct reader *reader;
    struct writer *writer;
};

struct socket *server_socket_accept(struct server_socket *serverSocket)
{
    int handle = socket_accept(serverSocket->handle);
    struct reader *reader = malloc(sizeof(struct reader));
    struct writer *writer = malloc(sizeof(struct writer));
    struct socket *socket = malloc(sizeof(struct socket));
    reader->handle = handle;
    reader->bufferStart = reader->buffer;
    reader->bufferEnd = reader->buffer;
    writer->handle = handle;
    socket->handle = handle;
    socket->reader = reader;
    socket->writer = writer;
    return socket;
}

struct socket *create_client_socket(int port)
{
    int handle = socket_create(port);
    struct reader *reader = malloc(sizeof(struct reader));
    struct writer *writer = malloc(sizeof(struct writer));
    struct socket *socket = malloc(sizeof(struct socket));
    reader->handle = handle;
    writer->handle = handle;
    socket->handle = handle;
    socket->reader = reader;
    socket->writer = writer;
    return socket;
}

struct reader *socket_get_reader(struct socket *socket)
{
    return socket->reader;
}

struct writer *socket_get_writer(struct socket *socket)
{
    return socket->writer;
}

void socket_close(struct socket *socket)
{
#ifdef WIN32
    closesocket(socket->handle);
#else
    close(socket->handle);
#endif
    free(socket->reader);
    free(socket->writer);
    free(socket);
}

File: tests/rust/unverified/platform/src/sockets.rs
#[cfg(windows)]
mod c {
    pub type CHAR = libc::c_char;

    #[repr(C)]
    #[cfg(any(target_arch = "aarch64", target_arch = "x86_64"))]
    pub struct WSADATA {
        pub wVersion: u16,
        pub wHighVersion: u16,
        pub iMaxSockets: u16,
        pub iMaxUdpDg: u16,
        pub lpVendorInfo: PSTR,
        pub szDescription: [u8; 257],
        pub szSystemStatus: [u8; 129],
    }

    pub type WSA_ERROR = i32;

    pub type ADDRESS_FAMILY = u16;
    pub const AF_INET: ADDRESS_FAMILY = 2u16;
    pub const PF_INET: ADDRESS_FAMILY = AF_INET;
    pub const SOCK_STREAM: i32 = 1;
    pub const IPPROTO_TCP: i32 = 6;

    pub type SOCKET = usize;
    pub const INVALID_SOCKET: SOCKET = -1i32 as _;

    #[repr(C)]
    #[derive(Copy, Clone)]
    pub struct in_addr {
        pub s_addr: u32,
    }

    pub const INADDR_ANY: u32 = 0;

    #[repr(C)]
    pub struct sockaddr {
        pub sa_family: ADDRESS_FAMILY,
        pub sa_data: [u8; 14],
    }

    #[repr(C)]
    #[derive(Copy, Clone)]
    pub struct sockaddr_in {
        pub sin_family: ADDRESS_FAMILY,
        pub sin_port: u16,
        pub sin_addr: in_addr,
        pub sin_zero: [CHAR; 8],
    }

    pub type socklen_t = c_int;

    pub type SEND_RECV_FLAGS = i32;

    #[link(name = "ws2_32")] 
    extern "system" { 
        pub fn WSAStartup(wversionrequested: u16, lpwsadata: *mut WSADATA) -> i32; 
        pub fn WSAGetLastError() -> WSA_ERROR;
        pub fn socket(af: i32, type_: i32, protocol: i32) -> SOCKET;
        pub fn bind(s: SOCKET, name: *const sockaddr, namelen: i32) -> i32;
        pub fn listen(s: SOCKET, backlog: i32) -> i32;
        pub fn accept(s: SOCKET, addr: *mut sockaddr, addrlen: *mut i32) -> SOCKET;
        pub fn recv(s: SOCKET, buf: PSTR, len: i32, flags: SEND_RECV_FLAGS) -> i32;
        pub fn send(s: SOCKET, buf: PCSTR, len: i32, flags: SEND_RECV_FLAGS) -> i32;
        pub fn closesocket(s: SOCKET) -> i32;
    }
}

#[cfg(unix)]
use libc as c;

#[cfg(windows)]
unsafe fn print_socket_error_message(api: &std::ffi::CStr) {
    println!("Socket API call '{}' failed: error code {}\n", api, c::WSAGetLastError());
}

#[cfg(unix)]
unsafe fn print_socket_error_message(api: &std::ffi::CStr) {
    libc::perror(api.as_ptr());
}

#[cfg(windows)]
type SOCKET = c::SOCKET;

#[cfg(windows)]
const INVALID_SOCKET: SOCKET = c::INVALID_SOCKET;

#[cfg(unix)]
type SOCKET = libc::c_int;

#[cfg(unix)]
const INVALID_SOCKET: SOCKET = -1;

#[derive(Copy, Clone)]
pub struct Socket {
    socket: SOCKET
}

impl Socket {

    #[cfg(windows)]
    unsafe fn initialize_sockets_api() {
        let windowsSocketsApiData: libc::WSADATA;
        libc::WSAStartup(libc::MAKEWORD(2, 0), &mut windowsSocketsApiData);
    }

    #[cfg(unix)]
    unsafe fn initialize_sockets_api() {}

    pub unsafe fn listen(port: u16) -> Socket {

        Self::initialize_sockets_api();

        let server_socket: SOCKET = c::socket(c::PF_INET, c::SOCK_STREAM, c::IPPROTO_TCP);
        if INVALID_SOCKET == server_socket {
            print_socket_error_message(c"socket()");
            std::process::abort();
        }
        let mut server_name: c::sockaddr_in = std::mem::zeroed();
        server_name.sin_addr.s_addr = u32::to_be(c::INADDR_ANY);
        server_name.sin_family = c::AF_INET.try_into().unwrap();
        server_name.sin_port = u16::to_be(port);
    
        {
            let status = c::bind(server_socket, &server_name as *const c::sockaddr_in as *const c::sockaddr, std::mem::size_of::<c::sockaddr_in>().try_into().unwrap());
            if INVALID_SOCKET == status {
                print_socket_error_message(c"bind()");
                std::process::abort();
            }
        }
    
        {
            let status = c::listen(server_socket, 5); // Allow 5 pending incoming connection requests
            if INVALID_SOCKET == status {
                print_socket_error_message(c"listen()");
                std::process::abort();
            }
        }
    
        return Socket { socket: server_socket };
    }
    
    pub unsafe fn accept(self) -> Socket {
        let mut client_name: c::sockaddr_in = std::mem::zeroed();
        let mut client_name_length: c::socklen_t = std::mem::size_of::<c::sockaddr_in>().try_into().unwrap();
        let client_socket = c::accept(self.socket, &mut client_name as *mut c::sockaddr_in as *mut c::sockaddr, &mut client_name_length as *mut c::socklen_t);
        if INVALID_SOCKET == client_socket {
            print_socket_error_message(c"accept()");
            std::process::abort();
        }
    
        return Socket { socket: client_socket };
    }

    pub unsafe fn receive(self, buffer: *mut u8, length: usize) -> usize {
        let result = c::recv(self.socket, buffer as *mut libc::c_void, length, 0);
        if result < 0 {
            print_socket_error_message(c"recv()");
            std::process::abort();
        }
        result.try_into().unwrap()
    }

    pub unsafe fn send(self, buffer: *const u8, length: usize) {
        if c::send(self.socket, buffer as *mut libc::c_void, length, 0) < 0 {
            // print_socket_error_message(c"send");
            // std::process::abort();
        }
    }

    #[cfg(windows)]
    pub unsafe fn close(self) {
        c::closesocket(self.socket);
    }

    #[cfg(unix)]
    pub unsafe fn close(self) {
        libc::close(self.socket);
    }
    
}
